mutable_batch_lp/src/lib.rs

```diff
@@ -139,38 +154,155 @@ pub fn lines_to_batches_stats(
     converter.finish()
 }
 
-/// Writes the [`ParsedLine`] to the [`MutableBatch`]
+/// An error applying an already-parsed line protocol line ([`ParsedLine`]) to a
+/// [`MutableBatch`].
+#[allow(missing_copy_implementations)]
+#[derive(Debug, Snafu)]
+pub enum LineWriteError {
+    /// A transparent error wrapper over the underling mutable batch error.
+    #[snafu(display("{}", source))]
+    MutableBatch {
+        /// The underlying error
+        source: mutable_batch::writer::Error,
+    },
+
+    /// The specified tag name appears twice in one LP line, with conflicting
+    /// values.
+    #[snafu(display(
+        "the tag '{}' is specified more than once with conflicting values",
+        name
+    ))]
+    DuplicateTag {
+        /// The duplicated tag name.
+        name: String,
+    },
+
+    /// The specified field name appears twice in one LP line, with conflicting
+    /// types.
+    #[snafu(display(
+        "the field '{}' is specified more than once with conflicting types",
+        name
+    ))]
+    ConflictedFieldTypes {
+        /// The duplicated field name.
+        name: String,
+    },
+}
+
+/// Writes the [`ParsedLine`] to the [`MutableBatch`], respecting the edge case
+/// semantics described in [`LinesConverter::write_lp()`].
 pub fn write_line(
     writer: &mut Writer<'_>,
     line: &ParsedLine<'_>,
     default_time: i64,
-) -> mutable_batch::writer::Result<()> {
-    for (tag_key, tag_value) in line.series.tag_set.iter().flatten() {
-        writer.write_tag(tag_key.as_str(), None, std::iter::once(tag_value.as_str()))?
+) -> Result<(), LineWriteError> {
+    // Only allocate the seen tags hashset if there are tags.
+    if let Some(tags) = &line.series.tag_set {
+        let mut seen = HashSet::with_capacity(tags.len());
+        for (tag_key, tag_value) in tags.iter() {
+            // Check if a field with this name has been observed previously.
+            if !seen.insert(tag_key) {
+                // This tag_key appears more than once, with differing values.
+                //
+                // This is always an error.
+                return Err(LineWriteError::DuplicateTag {
+                    name: tag_key.to_string(),
+                });
+            }
```

Comment:
Can use `snafu::ensure`

```suggestion
            // This tag_key appears more than once, with differing values.
            //
            // This is always an error.
            ensure!(seen.insert(tag_key), DuplicateTagSnafu {
                name: tag_key
            });
```

Reply:
maybe `ensure!(!seen` rather than `ensure!(seen` 🤔

Reply:
To be honest, I find this more confusing that the `if...return err()` but I guess I've not learnt snafu enough for it to be natural.

Reply:
I agree the current formulation (without `ensure!`) is quite clear and I don't think it needs to be changed